File: cpp/field.cpp
//----------------------------------------------------------------//
// gcafpp code by R. Saavedra. Email: saavestro@gmail.com         //
// For more information visit https://github.com/Saavestro/gcafpp //
//----------------------------------------------------------------//
#include "shared.h"

const double temp_prof(const double&);


void q_currents(const int &k){ // q profile and currents
	const double ew= // Inverse aspect ratio
		(double)field.i_rmin/field.i_rmaj;
	const double ra= // r position over minor radius
		sqrt(pol[k]/code.d_polw);
	const double rap= // ra pol derivative
		0.5*ra/pol[k];
	const double lmsq= // lambda squared
		(double)field.i_rmin*field.i_rmin*field.i_qaxi
		/(field.i_qwll -field.i_qaxi);

	qf[k]= // q profile
		(double)field.i_qaxi
		+field.i_qaxi*ra*ra*field.i_rmin*field.i_rmin/lmsq;
	qf_pol[k]= // q profile derivative
		field.i_qaxi*field.i_rmin*field.i_rmin/lmsq/code.d_polw;
	cg[k]= // poloidal current
		field.i_rmaj/(1.0 +ra*ew*cos(tht[k]))/code.d_leng;
	cg_pol[k]= // poloidal current derivative
		rap*ew*cos(tht[k])/(1.0 +ra*ew*cos(tht[k]))/code.d_leng;
	ci[k]= // toroidal current
		field.i_rmin*ra*ew/(1.0 +ra*ew*cos(tht[k]))/qf[k]/code.d_leng;
	const double ci_polft= rap*ew;
	const double ci_polst= -qf_pol[k]/qf[k]/qf[k];
	const double ci_poltt= rap*ew*cos(tht[k])/(1.0 +ra*ew*cos(tht[k]));
	ci_pol[k]= // toroidal current derivative
		ci_polft*ci_polst*ci_poltt/code.d_leng;
}


void field_tkmk(const int& k){ // tokamak magnetic field
	q_currents(k);
	const double ew= (double)field.i_rmin/field.i_rmaj;
	const double ra= sqrt(pol[k]/code.d_polw);
	const double rap= 0.5*ra/pol[k];
	const double bft= (double)field.i_baxi;
	const double bst= ra*ew*field.i_baxi/qf[k];
	const double bstp= rap*ew*field.i_baxi/qf[k]
		-ra*ew*field.i_baxi*qf_pol[k]/qf[k]/qf[k];

	bf[k]= // b field
		sqrt(bft*bft +bst*bst)/field.i_baxi/(1.0 +ew*ra*cos(tht[k]));
	const double bfpft= - ew*rap*cos(tht[k])/(1.0 +ew*ra*cos(tht[k]))
		/(1.0 +ew*ra*cos(tht[k]));
	const double bfpst= bst*bstp/sqrt(bft*bft +bst*bst);
	bf_pol[k]= // b field pol derivative
		bfpst/field.i_baxi/(1.0 +ew*ra*cos(tht[k]))
		+bfpft*sqrt(bft*bft +bst*bst)/field.i_baxi;
	bf_tht[k]= // b field tht derivative
		ew*ra*sin(tht[k])*sqrt(bft*bft +bst*bst)
		/field.i_baxi/(1.0 +ew*ra*cos(tht[k]))/(1.0 +ew*ra*cos(tht[k]));
	bf_zet[k]= // b field zet derivative
		0.0;
}


void electrostatic_potential(const int& k){ // electrostatic potential
	if(field.potp==POT_CONS){
		const double dum= pol[k]/code.d_polw;
		pt[k]= code.d_pt*pol[k]*pol[k]/code.d_polw/code.d_polw;
		pt_pol[k]= code.d_pt*pol[k]/code.d_polw/code.d_polw;
		pt_tht[k]= 0.0;
		pt_zet[k]= 0.0;
	}
	else{ // needs to be fixed
		const double tmpdum= KEV_TO_ERG*species[0].i_temp/code.d_enax;
		pt[k]= code.d_pt +tmpdum*temp_prof(pol[k]/code.d_polw);
		pt_pol[k]= code.d_pt
			-tmpdum*temp_prof(pol[k]/code.d_polw)/code.d_polw;
		pt_tht[k]= 0.0;
		pt_zet[k]= 0.0;
	}
}


void field_stel(const int& k){ // stellarator magnetic field
	q_currents(k);
	const int lc= 2;
	const int mc= 6;
	const double sg= -1;
	const double ew0= (double)field.i_rmin/field.i_rmaj;
	const double ew1= 0.4;
	const double ewm= 0.4;
	const double ewp= 0.4;
	const double pfrs0= sqrt(pol[k]/code.d_polw);
	const double pfrs1= pow(pfrs0,lc);
	const double pfrsm= pow(pfrs0,lc-1);
	const double pfrsp= pow(pfrs0,lc+1);
	const double pfrs0p= 0.5*pfrs0/pol[k];
	const double pfrs1p= 0.5*lc*pfrs1/pol[k];
	const double pfrsmp= 0.5*(lc -1)*pfrsm/pol[k];
	const double pfrspp= 0.5*(lc +1)*pfrsp/pol[k];
	const double coss= cos(tht[k]);
	const double sins= sin(tht[k]);
	const double cosl= cos(lc*tht[k] -mc*zet[k]);
	const double sinl= sin(lc*tht[k] -mc*zet[k]);
	const double coslm= cos((lc -1)*tht[k] -mc*zet[k]);
	const double sinlm= sin((lc -1)*tht[k] -mc*zet[k]);
	const double coslp= cos((lc +1)*tht[k] -mc*zet[k]);
	const double sinlp= sin((lc +1)*tht[k] -mc*zet[k]);

	bf[k]= 1.0 -ew0*pfrs0*coss -ew1*pfrs1*cosl
		+0.5*sg*(ewm*pfrsm*coslm +ewp*pfrsp*coslp);
	bf_pol[k]= -ew0*pfrs0p*coss -ew1*pfrs1p*cosl
		+0.5*sg*(ewm*pfrsmp*coslm +ewp*pfrspp*coslp);
	bf_tht[k]= ew0*pfrs0*sins +lc*ew1*pfrs1*sinl
		-0.5*sg*((lc -1)*ewm*pfrsm*sinlm +(lc +1)*ewp*pfrsp*sinlp);
	bf_zet[k]= -mc*ew1*pfrs1*sinl
		+0.5*sg*(mc*ewm*pfrsm*sinlm +mc*ewp*pfrsp*sinlp);
}


void perturbation(const int& k){ // magnetic perturbation
	const double ph= (double)field.i_phas*M_PI/180;
	const int mi= field.i_mmod;
	const int mf= field.i_mmod;
	const int n= field.i_nmod;
	const int rp= n;
	for(int m=mi;m<=mf;m++){
		double facx= (double)m/(m +rp);
		double ra= field.d_ampx*pow(pol[k]/code.d_polw/facx,m)
			*pow(1.0 -pol[k]/code.d_polw,rp)/pow(1.0 -facx,rp);
		double rap= m*ra/pol[k] -rp*ra/code.d_polw/(1.0 -pol[k]/code.d_polw);
		double sin_long= sin(n*zet[k] -m*tht[k]
				+ph -code.d_pert_freq*tim[k]);
		double cos_long= cos(n*zet[k] -m*tht[k]
				+ph -code.d_pert_freq*tim[k]);
		al[k]= // alpha perturbation
			ra*sin_long;
		al_pol[k]= // alpha pol derivative
			rap*sin_long;
		al_tht[k]= // alpha tht derivative
			m*ra*cos_long;
		al_zet[k]= // alpha zet derivative
			n*ra*cos_long;
		al_tim[k]= // alpha tim derivative
			code.d_pert_freq*ra*cos_long;
	}
}

File: fortran/src/field.f90
!---------------------------------------------------------------------!
! GCAF= Guiding Center Analytic Field code by R. Saavedra             !
! For more information please visit https://git.rsaavedra.xyz/?p=gcaf.git !
!---------------------------------------------------------------------!
!---------------------------------------------------------!
subroutine field(k)
!-----------------!
! External fields !
!-----------------!
!=========================================================!
	use shared,only:&
		keverg,& !keV to erg
		enaxo,&  !on axis en
		npert,&  !perturbation on/off
		nepot,&  !input pt profile pres/flat
		leno,&   !characteristic length
		ntmpi,&  !ions temp in keV
		ntmpe,&  !elec temp in keV
		pol,&    !poloidal flux
		polw,&   !on wall poloidal flux
		tht,&    !theta
		zet,&    !zeta
		rmin,&   !minor radius
		rmaj,&   !major radius
		cg,&     !poloidal current
		cgp,&    !g pol derivative
		ci,&     !toroidal current
		cip,&    !ri pol derivative
		qf,&     !q profile
		bf,&     !b field magnitude
		bm,&     !on axis b field magnitude
		dbdp,&   !b pol derivative
		dbdt,&   !b tht derivative
		dbdz,&   !b zet derivative
		pt,&     !electrostatic potential
		pto,&    !pt magnitude
		dptdp,&  !pt pol derivative
		dptdt,&  !pt tht derivative
		dptdz    !pt zet derivative
	implicit none
	integer(8):: k
	integer(8):: tmprat
	real(8):: pdum      !dummy pol
	real(8):: tdum      !dummy tht
	real(8):: zdum      !dummy zet
	real(8):: temp_prof !temp_prof function
	real(8):: tmpdum    !dummy temp
	real(8):: ew        !inverse aspect ratio
	real(8):: q0   		!on axis q profile
	real(8):: qw   		!on wall q profile
	real(8):: ra   		!r/a
	real(8):: rap  		!r/a pol derivative
	real(8):: lmsq 		!lambda squared
	real(8):: bfpft 	!b first term pol derivative
	real(8):: bfpst 	!b second term pol derivative
	real(8):: bt   		!b toroidal
	real(8):: bp   		!b poloidal
	real(8):: bpp  		!bp pol derivative
	real(8):: qfp  		!q pol derivative
	real(8):: dum
	real(8):: cipft
	real(8):: cipst
	real(8):: ciptt
!=========================================================!

	pdum= pol(k)
	tdum= tht(k)
	zdum= zet(k)
	ew= 1.d0*rmin/rmaj
	qw= 5.d0
	q0= 1.d0
	ra= sqrt(pdum/polw)
	rap= .5d0*ra/pdum
	lmsq= rmin*rmin*q0/(qw -q0)
	qf(k)= q0 +q0*ra*ra*rmin*rmin/lmsq
	qfp= q0*rmin*rmin/lmsq/polw
	cg(k)= rmaj/(1.d0 +ra*ew*cos(tdum))/leno/bm
	cgp(k)= rap*ew*cos(tdum)/(1.d0 +ra*ew*cos(tdum))
	ci(k)= rmin*ra*ew/(1.d0 +ra*ew*cos(tdum))/qf(k)/leno/bm
	cipft= rap*ew
	cipst= -qfp/qf(k)/qf(k)
	ciptt= rap*ew*cos(tdum)/(1.d0 +ra*ew*cos(tdum))
	cip(k)= cipft*cipst*ciptt

	!print *, cg(k)*leno*bm

	bt= bm
	bp= ra*ew*bm/qf(k)
	bpp= ew*bm*rap/qf(k) -ew*bm*ra*qfp/qf(k)/qf(k)
	bfpft= - ew*rap*cos(tdum)/(1.d0 +ew*ra*cos(tdum)) &
		/(1.d0 +ew*ra*cos(tdum))
	bfpst= bp*bpp/sqrt(bt*bt +bp*bp)
	bf(k)= sqrt(bt*bt +bp*bp)/bm/(1.d0 +ew*ra*cos(tdum))
	dbdp(k)= bfpst/bm/(1.d0 +ew*ra*cos(tdum)) &
		+bfpft*sqrt(bt*bt +bp*bp)/bm
	dbdt(k)= ew*ra*sin(tdum)*sqrt(bt*bt +bp*bp)/bm &
		/(1.d0 +ew*ra*cos(tdum))/(1.d0 +ew*ra*cos(tdum))
	dbdz(k)= 0.d0
	!simpler b
	!bf(k)= 1.d0 -ew*ra*cos(tdum)
	!dbdp(k)= -ew*rap*cos(tdum)
	!dbdt(k)= ew*ra*sin(tdum)
	!dbdz(k)= 0.d0

	pt(k)= 0.d0
	dptdp(k)= 0.d0
	dptdt(k)= 0.d0
	dptdz(k)= 0.d0
	dum= pdum/polw
	tmprat= ntmpi/ntmpi
	if(nepot.eq.0)then
		pt(k)= pto*dum*dum
		dptdp(k)= pto*dum/polw
		dptdt(k)= 0.d0
		dptdz(k)= 0.d0
	endif
	if(nepot.eq.1)then
		tmpdum= keverg/enaxo
		pt(k)= pto +tmpdum*temp_prof(dum,tmprat)
		dptdp(k)= pto -tmpdum*temp_prof(dum,tmprat)/polw
		dptdt(k)= 0.d0
		dptdz(k)= 0.d0
	endif
	if(npert.eq.1) call ptrba(k)
	if(pdum.ne.pdum) then
		write(0,'(a18)') 'POLOIDAL FLUX NaN'
		stop
	endif
	return
end subroutine
!---------------------------------------------------------!


!---------------------------------------------------------!
subroutine ptrba(k)
!--------------------------------!
! Mynick alfa perturbation model !
! Phys. Fluids B 5, 1471 (1993)  !
!--------------------------------!
!=========================================================!
	use shared,only: &
		pi,&     !pi number
		namx,&   !input perturbation amplitude
		time,&   !time
		pol,&    !poloidal flux
		tht,&    !theta
		zet,&    !zeta
		al,&     !alfa perturbation
		dadp,&   !al pol derivative
		dadt,&   !al tht derivative
		dadz,&   !al zet derivative
		dadtim,& !al time derivative
		prtfrq,& !input perturbation frequency
		mmod,&   !input m mode number
		nmod,&   !input n mode number
		nphs,&	 !input perturbation phase
		polw     !on wall poloidal flux
	implicit none
	integer(8):: m   !m mode number
	integer(8):: n   !n mode number
	integer(8):: k   !particle index
	integer(8):: mi  !initial m mode
	integer(8):: mf  !final m mode
	real(8):: pdum   !pol dummy
	real(8):: tdum   !tht dummy
	real(8):: zdum   !zet dummy
	real(8):: ph     !phase
	real(8):: facx   !rmx/a
	real(8):: amx    !scaled namx
	real(8):: ra     !amp
	real(8):: rap    !amp prime
	real(8):: sn     !sine
	real(8):: cs     !cosine
	real(8):: rp     !Mynick model p=n for qnm>1
	real(8):: rotfrq !island rotation frequency
!=========================================================!
	al(k)= 0.d0
	dadp(k)= 0.d0
	dadt(k)= 0.d0
	dadz(k)= 0.d0
	dadtim(k)= 0.d0
	pdum= pol(k)
	tdum= tht(k)
	zdum= zet(k)

	ph= nphs*pi/180
	amx= namx
	mi=mmod
	mf=mmod
	n=nmod
	rp= 2 !n
	do m=mi,mf
		facx= 1.d0*m/(m +rp)
		ra= amx*(pdum/polw/facx)**m*(1.d0 -pdum/polw)**rp &
			/(1.d0 -facx)**rp
		rap= m*ra/pdum -rp*ra/polw/(1.d0 -pdum/polw)
		sn= sin(n*zdum -m*tdum +ph -prtfrq*time(k))
		cs= cos(n*zdum -m*tdum +ph -prtfrq*time(k))
		al(k)= al(k) +ra*sn
		dadp(k)= dadp(k) +rap*sn
		dadt(k)= dadt(k) -m*ra*cs
		dadz(k)= dadz(k) +n*ra*cs
		dadtim(k)= dadtim(k) -prtfrq*ra*cs
	enddo
	return
end subroutine
!---------------------------------------------------------!
